server/issue.go

```diff
@@ -882,40 +882,137 @@ func (s *Server) getPipelineCreateForDatabaseSchemaUpdateGhost(ctx context.Conte
+	if int64(databaseIDCount) > maximumTaskLimit {
+		return nil, echo.NewHTTPError(http.StatusForbidden, fmt.Sprintf("Current plan can update up to %d databases, got %d.", maximumTaskLimit, databaseIDCount))
+	}
+
+	// aggregatedMatrix is the aggregated matrix by deployments.
+	// databaseToMigrationList is the mapping from database ID to migration detail.
+	aggregatedMatrix := make([][]*api.Database, len(deploySchedule.Deployments))
+	databaseToMigrationList := make(map[int][]*api.UpdateSchemaGhostDetail)
+
+	dbList, err := s.store.FindDatabase(ctx, &api.DatabaseFind{
+		ProjectID: &issueCreate.ProjectID,
+	})
+	if err != nil {
+		return nil, echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch databases in project ID: %v", issueCreate.ProjectID)).SetInternal(err)
+	}
 
-		database, err := s.store.GetDatabase(ctx, &api.DatabaseFind{ID: &detail.DatabaseID})
+	if databaseIDCount == 0 {
+		// Deploy to all tenant databases.
+		migrationDetail := c.DetailList[0]
+		matrix, err := utils.GetDatabaseMatrixFromDeploymentSchedule(deploySchedule, dbList)
 		if err != nil {
-			return nil, echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("failed to fetch database ID: %v", detail.DatabaseID)).SetInternal(err)
+			return nil, echo.NewHTTPError(http.StatusInternalServerError, "Failed to build deployment pipeline").SetInternal(err)
 		}
-		if database == nil {
-			return nil, echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("database ID not found: %d", detail.DatabaseID))
+		aggregatedMatrix = matrix
+		for _, databaseList := range matrix {
+			for _, database := range databaseList {
+				// There should be only one migration per database for tenant mode deployment.
+				databaseToMigrationList[database.ID] = []*api.UpdateSchemaGhostDetail{migrationDetail}
+			}
 		}
+	} else {
+		databaseMap := make(map[int]*api.Database)
+		for _, db := range dbList {
+			databaseMap[db.ID] = db
+		}
+		for _, d := range c.DetailList {
+			database, ok := databaseMap[d.DatabaseID]
+			if !ok {
+				return nil, echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Database ID %d not found in project %d", d.DatabaseID, issueCreate.ProjectID))
+			}
+			matrix, err := utils.GetDatabaseMatrixFromDeploymentSchedule(deploySchedule, []*api.Database{database})
+			if err != nil {
+				return nil, echo.NewHTTPError(http.StatusInternalServerError, "Failed to build deployment pipeline").SetInternal(err)
+			}
+			for i, databaseList := range matrix {
+				aggregatedMatrix[i] = append(aggregatedMatrix[i], databaseList...)
+			}
+			databaseToMigrationList[database.ID] = append(databaseToMigrationList[database.ID], d)
+		}
+	}
 
-		taskCreateList, taskIndexDAGList, err := createGhostTaskList(database, c.VCSPushEvent, detail, schemaVersion)
+	create := &api.PipelineCreate{
+		Name:      "Update database schema (gh-ost) pipeline",
+		CreatorID: issueCreate.CreatorID,
+	}
+
+	for i, databaseList := range aggregatedMatrix {
```

Comment:
Is this something that we can share with none gh-ost DDL/DML code? A lot of things are shared I think.